src/self_driving/planning/seq_tuning/sequence_tuning.cpp

```diff
@@ -0,0 +1,212 @@
+#include "self_driving/planning/seq_tuning/sequence_tuning.h"
+
+#include <map>
+#include <vector>
+
+#include "common/managed_pointer.h"
+#include "loggers/selfdriving_logger.h"
+#include "planner/plannodes/abstract_plan_node.h"
+#include "self_driving/planning/action/generators/change_knob_action_generator.h"
+#include "self_driving/planning/action/generators/index_action_generator.h"
+#include "self_driving/planning/pilot_util.h"
+#include "self_driving/planning/seq_tuning/graph_solver.h"
+#include "transaction/transaction_manager.h"
+
+namespace noisepage::selfdriving::pilot {
+
+SequenceTuning::SequenceTuning(common::ManagedPointer<Pilot> pilot,
+                               common::ManagedPointer<selfdriving::WorkloadForecast> forecast,
+                               uint64_t end_segment_index)
+    : pilot_(pilot), forecast_(forecast), end_segment_index_(end_segment_index) {
+  transaction::TransactionContext *txn = pilot->txn_manager_->BeginTransaction();
+
+  std::vector<std::unique_ptr<planner::AbstractPlanNode>> plans;
+  // vector of query plans that the search tree is responsible for
+  PilotUtil::GetQueryPlans(pilot, common::ManagedPointer(forecast_), end_segment_index, txn, &plans);
+
+  std::vector<action_id_t> candidate_actions;
+  // populate structure_map_, candidate_structures_
+  IndexActionGenerator().GenerateActions(plans, pilot->settings_manager_, &structure_map_, &candidate_actions);
+
+  for (auto &action : candidate_actions)
+    if (structure_map_.at(action)->GetSQLCommand().rfind("create index", 0) == 0) {
+      candidate_structures_.emplace_back(action);
+      SELFDRIVING_LOG_DEBUG("Candidate structure: ID {} Command {}", action,
+                            structure_map_.at(action)->GetSQLCommand());
+    }
+
+  for (const auto &it UNUSED_ATTRIBUTE : structure_map_) {
+    SELFDRIVING_LOG_DEBUG("Generated action: ID {} Command {}", it.first, it.second->GetSQLCommand());
+  }
+
+  pilot->txn_manager_->Abort(txn);
+
+  std::vector<double> default_segment_cost;
+  // first compute the cost of each segment when no action is applied
+  for (uint64_t segment_index = 0; segment_index <= end_segment_index_; segment_index++) {
+    default_segment_cost.emplace_back(PilotUtil::ComputeCost(pilot_, forecast_, segment_index, segment_index));
+  }
+  default_segment_cost_ = default_segment_cost;
+}
+
+void SequenceTuning::BestAction(
+    std::vector<std::set<std::pair<const std::string, catalog::db_oid_t>>> *best_actions_seq,
+    uint64_t memory_constraint) {
+  // cost-based pruning for each individual action
+  std::map<action_id_t, std::tuple<std::vector<std::set<action_id_t>>, std::set<std::set<action_id_t>>, double>>
```

Comment:
I think it would be clearer if you define a `struct` for this tuple with the three variables defined separately.